### PR TITLE
CLDR-14989 GitHub Action to build ICU with CLDR

### PR DESCRIPTION
**Pull request by srl295**
CLDR-14989

- [X] This PR completes the ticket.

### CODE REVIEW

File: .github/workflows/build-icu.yml [hunk 1]
```diff
@@ -63,10 +63,82 @@ jobs:
           mvn  -s .github/workflows/mvn-settings.xml -B  -DCLDR_DIR=$(pwd) -DCLDR_GITHUB_ANNOTATIONS=true --file=tools/pom.xml -pl cldr-code  \
             exec:java -Dexec.mainClass=org.unicode.cldr.tool.GenerateProductionData \
             -Dexec.args="-d target/cldr-prod"
+          (cd target/cldr-prod/ && zip -r ../cldr-prod.zip *)
```
Comment (srl295) [inline .github/workflows/build-icu.yml:66]: seems to be more performant to zip up first

**Comment by srl295** [comment]
https://github.com/srl295/cldr/actions/runs/1211247730

**Comment by srl295** [comment]
Generates these artifacts:
- cldr-code:  the main CLDR jar
- cldr-prod: Production CLDR xml data (aka cldr-staging/production)
- icu-data: a .zip file roughly like the icu-data.zip, with built ICU data
- icu4c:  distribution of ICU4C built w/ clang on ubuntu
- icu4j: distribution of ICU4J built w/ OpenJDK16

The latter two targets were copied from icu's

**Comment by macchiati** [comment]
Thanks!

It isn't clear from https://github.com/srl295/cldr/actions/runs/1211247730
where the artifacts are stored for download. That is, although the CLDR
data and ICU data re built, I don't see where to fetch them from.

Mark

**Comment by srl295** [comment]
> make: *** No rule to make target 'out/build/icudt70l/uprops.icu', needed by 'out/icu4j/icudata.jar'.  Stop.

Any ideas @yumaoka ?

**Comment by srl295** [comment]
> Thanks!
> 
> It isn't clear from https://github.com/srl295/cldr/actions/runs/1211247730
> where the artifacts are stored for download. That is, although the CLDR
> data and ICU data re built, I don't see where to fetch them from.
> 

They are attached to the action run .. may not show up since this is a fork. 

Try the action in CLDR.

**Comment by srl295** [comment]
Ready to merge.

**Comment by srl295** [comment]
It's at the bottom of https://github.com/srl295/cldr/actions/runs/1211247730 as shown below. If it doesn't show it may be because it's in my own fork and won't give access.

image

File: .github/workflows/build-icu.yml [hunk 2]
```diff
@@ -59,14 +59,194 @@ jobs:
...
+        run: >
+          env
+          TOOLS_ROOT=$(pwd)/icu/tools/
+          CLDR_DIR=$(pwd)/cldr
+          CLDR_DATA_DIR=$(pwd)/cldr-prod/
+          ant -f icu/tools/cldr/cldr-to-icu/build-icu-data.xml clean all && ln icu/icu4c/LICENSE icu/icu4c/source/data/
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - name: Zip up icu-data.zip
+        run: 'cd icu/icu4c/source && zip -rlq ../../../icu-data.zip data'
```
Comment (jefgen) [inline .github/workflows/build-icu.yml:155]: I was wondering why this created a double zip file (icu-data.zip inside icu-data.zip)... but it looks like this is a limitation of the GitHub Action:

https://github.com/actions/upload-artifact
>  One of the consequences of this limitation is that if a zip is uploaded during a workflow run and then downloaded from the UI, there will be a **double zip created**.

Oh well...  ¯\\_ (ツ)_/¯

File: .github/workflows/build-icu.yml [hunk 3]
```diff
@@ -59,14 +59,194 @@ jobs:
...
+        # Do this after restoring cache.
+        run: >
+          mvn -B install:install-file   -Dproject.parent.relativePath=""   -DgroupId=org.unicode.cldr
+          -DartifactId=cldr-api   -Dversion=0.1-SNAPSHOT   -Dpackaging=jar   -DgeneratePom=true
+          -Dfile=cldr-code/cldr-code.jar
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - name: Generate ICU Data
+        run: >
+          env
```
Comment (jefgen) [inline .github/workflows/build-icu.yml:147]: In the [cldr-icu-readme.txt](https://github.com/unicode-org/icu/blob/main/icu4c/source/data/cldr-icu-readme.txt#L125) file I noticed that it sets other environment vars, like `ANT_OPTS="-Xmx4096m"` and `ICU4C_DIR`/`ICU4J_ROOT`.

I guess they might not actually be needed?

(FWIW, I think the `ICU4C_DIR` will be auto-set [here](https://github.com/unicode-org/icu/blob/7851f708a3dd43fc1e0e9dd784e7dfe9a70915b2/icu4c/source/data/build.xml#L25), but I'm not sure about `ICU4J_ROOT`)

File: .github/workflows/build-icu.yml [hunk 4]
```diff
@@ -59,14 +59,194 @@ jobs:
...
+          name: icu-data
+          path: icu-data.zip
+  # clang release build with some options to enforce useful constraints.
+  # Includes dependency checker on an in-source, optimized build.
+  # Includes checking @draft etc. API tags vs. ifndef guards like
+  # U_HIDE_DRAFT_API and U_FORCE_HIDE_DRAFT_API.
+  # (FORCE guards make this tool pass but won't compile to working code.
+  # See the testtagsguards.sh script for details.)
+  icu4c-clang:
+    runs-on: ubuntu-latest
```
Comment (jefgen) [inline .github/workflows/build-icu.yml:168]: Wonder if maybe giving it a name, perhaps something like (?):
`    name: Build and test ICU4C, with new data`

**Review by jefgen — COMMENTED** [review]
This is really neat @srl295! 😊

File: .github/workflows/build-icu.yml (hunk 2 above)
Comment (srl295) [reply .github/workflows/build-icu.yml:155]: @jefgen it seemed to be quicker to upload a zip rather than to upload a 'directory'.

Comment (jefgen) [reply .github/workflows/build-icu.yml:155]: Ah, good point. 👍

File: .github/workflows/build-icu.yml (hunk 3 above)
Comment (srl295) [reply .github/workflows/build-icu.yml:147]: Didn't seem to be needed.

Comment (srl295) [reply .github/workflows/build-icu.yml:147]: `ICU4J_ROOT` is for the icu4j data install step (C Makefile) and is set on line 230, but is not needed here on line 147 - this is cldr-to-icu (David's tool)

Comment (jefgen) [reply .github/workflows/build-icu.yml:147]: Ah, sorry for missing that. Thanks!

**Comment by srl295** [comment]
New run: https://github.com/srl295/cldr/actions/runs/1222657915

**Review by jefgen — APPROVED** [review]

**Pull request merged**